File: scm-java/src/main/scheme/primitives/PrimitiveJiffy.java
package scheme.primitives;

import scheme.Primitive;
import scheme.SourcePos;

public class PrimitiveJiffy extends Primitive {
    @Override
    public String name() {
        return "%jiffy";
    }

    @Override
    public String info() {
        return "Syntax: (%jiffy)\n" +
               "Library: (scheme time)\n" +
               "Description: Internal primitive. Returns the number of microseconds elapsed since the Unix epoch (1970-01-01 00:00:00 UTC).\n" +
               "Example:\n" +
               "  (%jiffy) => 1700000000000000";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 0, 0);
        var now = java.time.Instant.now();
        return now.getEpochSecond() * 1_000_000L + now.getNano() / 1000;
    }
}

File: scm-csharp/scheme/primitives/PrimitiveJiffy.cs
namespace scheme;

public class PrimitiveJiffy : Primitive
{
    private static readonly long UnixEpochMicroseconds =
        new DateTimeOffset(1970, 1, 1, 0, 0, 0, TimeSpan.Zero).Ticks / 10L;

    public override string Name() => "%jiffy";

    public override string Info() =>
        "Syntax: (%jiffy)\n" +
        "Library: (scm core)\n" +
        "Description: Internal primitive. Returns the number of microseconds elapsed since the Unix epoch (1970-01-01 00:00:00 UTC).\n" +
        "Example:\n" +
        "  (%jiffy) => 1700000000000000";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 0, 0);
        return DateTimeOffset.UtcNow.Ticks / 10L - UnixEpochMicroseconds;
    }
}
